dali/test/dali_operator_test.h

```diff
@@ -23,30 +23,198 @@
 #include <vector>
 #include <memory>
 #include <map>
+#include <string>
+#include <utility>
+#include "dali/test/operator_argument.h"
+
 
 namespace dali {
 namespace testing {
 
-using Arguments = std::map<ArgumentKey, double>;  // TODO(mszolucha) some generalization boost::any?
+using Arguments = std::map<ArgumentKey, TestOpArg>;
+
+namespace detail {
+
+template<typename Backend>
+std::string BackendStringName() {
+  DALI_ENFORCE(false, "Backend not supported. You may want to write your own specialization");
+}
+
+
+template<>
+inline std::string BackendStringName<CPUBackend>() {
+  return "cpu";
+}
+
+
+template<>
+inline std::string BackendStringName<GPUBackend>() {
+  return "gpu";
+}
+
+
+}  // namespace detail
+
+inline std::unique_ptr<Pipeline> CreatePipeline(size_t batch_size, size_t num_threads) {
+  return std::unique_ptr<Pipeline>(
+          new Pipeline(static_cast<int>(batch_size), static_cast<int>(num_threads), 0));
+}
+
+
+inline void AddInputToPipeline(Pipeline &pipeline, const TensorListWrapper &input) {
+  assert(input && input.has_cpu());  // External input works only for CPUBackend
+  const std::string input_name = "input";
+  pipeline.AddExternalInput(input_name);
+  pipeline.SetExternalInput(input_name, *input.get<CPUBackend>());
+}
+
+
+inline void AddOperatorToPipeline(Pipeline &pipeline, const OpSpec &op_spec) {
+  pipeline.AddOperator(op_spec);
+}
+
+
+inline DeviceWorkspace CreateWorkspace() {
+  DeviceWorkspace ws;
+  return ws;
+}
+
+
+inline void RunPipeline(Pipeline &pipeline) {
+  pipeline.RunCPU();
+  pipeline.RunGPU();
+}
+
 
+template<typename OutputBackend>
+inline std::vector<TensorListWrapper>
+GetOutputsFromPipeline(Pipeline &pipeline, const std::string &output_backend) {
+  std::vector<TensorListWrapper> ret;
+  auto workspace = CreateWorkspace();
+  pipeline.Outputs(&workspace);
+  for (int output_idx = 0; output_idx < workspace.NumOutput(); output_idx++) {
+    ret.emplace_back(&workspace.template Output<OutputBackend>(output_idx));
+  }
+  return ret;
```

Comment:
You return a pointers to things in `workspace` that is freed after this return.

Reply:
Are you 100% sure of this? I skimmed through the code and I thought that pointers in this workspace are shared with the ones in executors tensor list pool, where they are actually created, so they should live as long as there is a pipeline, so till the end of `RunTest`. I think this issue might happen when `RunTest` is not the last call in `TEST_P` macro, because the outputs are returned through reference to `TensorListWrapper`. Quite possibly I'm missing something, so thanks in advance for clarification.

Reply:
Yeah, they should live as long as pipeline IIRC, but still, that kind of return seems prone to errors.